sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/binaryfile/BinaryFileFormat.scala

```diff
@@ -101,43 +98,27 @@ class BinaryFileFormat extends FileFormat with DataSourceRegister {
     val filterFuncs = filters.map(filter => createFilterFunction(filter))
 
     file: PartitionedFile => {
-      val path = file.filePath
-      val fsPath = new Path(path)
-
+      val path = new Path(file.filePath)
       // TODO: Improve performance here: each file will recompile the glob pattern here.
-      if (pathGlobPattern.forall(new GlobFilter(_).accept(fsPath))) {
-        val fs = fsPath.getFileSystem(broadcastedHadoopConf.value.value)
-        val fileStatus = fs.getFileStatus(fsPath)
-        val length = fileStatus.getLen
-        val modificationTime = fileStatus.getModificationTime
-
-        if (filterFuncs.forall(_.apply(fileStatus))) {
-          val stream = fs.open(fsPath)
-          val content = try {
-            ByteStreams.toByteArray(stream)
-          } finally {
-            Closeables.close(stream, true)
-          }
-
-          val fullOutput = dataSchema.map { f =>
-            AttributeReference(f.name, f.dataType, f.nullable, f.metadata)()
-          }
-          val requiredOutput = fullOutput.filter { a =>
-            requiredSchema.fieldNames.contains(a.name)
+      if (pathGlobPattern.forall(new GlobFilter(_).accept(path))) {
+        val fs = path.getFileSystem(broadcastedHadoopConf.value.value)
+        val status = fs.getFileStatus(path)
+        if (filterFuncs.forall(_.apply(status))) {
+          val values = requiredSchema.fieldNames.map {
+            case PATH => UTF8String.fromString(status.getPath.toString)
+            case LENGTH => status.getLen
+            case MODIFICATION_TIME => DateTimeUtils.fromMillis(status.getModificationTime)
+            case CONTENT =>
+              val stream = fs.open(status.getPath)
+              try {
+                ByteStreams.toByteArray(stream)
+              } finally {
+                Closeables.close(stream, true)
+              }
+            case other =>
+              throw new RuntimeException(s"Unsupported field name: ${other}")
           }
-
-          // TODO: Add column pruning
-          // currently it still read the file content even if content column is not required.
-          val requiredColumns = GenerateUnsafeProjection.generate(requiredOutput, fullOutput)
-
-          val internalRow = InternalRow(
-            UTF8String.fromString(path),
-            DateTimeUtils.fromMillis(modificationTime),
-            length,
-            content
-          )
-
-          Iterator(requiredColumns(internalRow))
+          Iterator.single(InternalRow(values: _*))
```

Comment:
Why don't project to unsafe row like previously?

Reply:
Does projecting to unsafe row improve performance ?

Reply:
Seems `dataSchema` is not used, is it possible that the required schema contains fields that not exist in the `dataSchema`?

Reply:
I recall that we return unsafe row if possible.

Reply:
Agree with @viirya, unsafe row can be more efficient in space, and if there are no gain to do this change, we'd better keep it unchanged.

Reply:
Updated using `UnsafeRowWriter`